src/pybind/mgr/cephadm/upgrade.py

```diff
@@ -349,6 +349,46 @@ def _prepare_for_mds_upgrade(
 
         return continue_upgrade
 
+    def _post_upgrade_redeploys(self):
+        # type: () -> None
+        daemons = self.mgr.cache.get_daemons()
+
+        # redeploy all daemons who are not redployed during upgrade
+        # and mgrs which could have been deployed by a mgr running old code
+        for d in daemons:
+            assert d.daemon_type is not None
+            assert d.daemon_id is not None
+            assert d.hostname is not None
+
+            if d.daemon_type in CEPH_UPGRADE_ORDER and d.daemon_type != 'mgr':
+                continue
+
+            # current active mgr cannot redeploy itself. Schedule it for redeploy
+            # later then continue redeploying other daemons.
+            if self.mgr.daemon_is_self(d.daemon_type, d.daemon_id):
+                self.mgr.cache.schedule_daemon_action(d.hostname, d.name(), 'redeploy')
+                continue
+
+            if not self._wait_for_ok_to_stop(d):
+                return
+
+            try:
+                self.mgr._daemon_action(
+                    d.daemon_type,
+                    d.daemon_id,
+                    d.hostname,
+                    'redeploy'
+                )
+            except Exception as e:
+                self._fail_upgrade('POST_UPGRADE_REDEPLOY_DAEMON', {
+                    'severity': 'warning',
+                    'summary': f'Redeploying daemon {d.name()} on host {d.hostname} after upgrade failed.',
+                    'count': 1,
+                    'detail': [
+                        f'Redeploy daemon: {d.name()}: {e}'
+                    ],
+                })
```

Comment:
and then you're returning from `_post_upgrade_redeploys` and then `_do_upgrade` cleans up the upgrade process and no user will ever see this error message